cs/azure/Trainer/LearnEventProcessorFactory.cs

```diff
@@ -0,0 +1,247 @@
+﻿using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ServiceBus.Messaging;
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using VW;
+using Newtonsoft.Json;
+using System.IO;
+using System.Diagnostics;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+using VW.Labels;
+using System.Text;
+using VowpalWabbit.Azure.Trainer;
+using VowpalWabbit.Azure.Trainer.Operations;
+using VW.Serializer;
+
+namespace VowpalWabbit.Azure
+{
+    internal sealed class TrainEventProcessorFactory : IEventProcessorFactory, IDisposable
+    {
+        private readonly TelemetryClient telemetry;
+        private readonly PerformanceCounters performanceCounters;
+        private readonly Learner trainer;
+
+        private EvalOperation evalOperation;
+        //private LatencyOperation latencyOperation;
+
+        private TransformManyBlock<PipelineData, PipelineData> deserializeBlock;
+        private TransformManyBlock<object, object> learnBlock;
+        private ActionBlock<object> checkpointBlock;
+
+        internal TrainEventProcessorFactory(OnlineTrainerSettingsInternal settings, Learner trainer, PerformanceCounters performanceCounters)
+        {
+            if (settings == null)
+                throw new ArgumentNullException(nameof(settings));
+
+            if (trainer == null)
+                throw new ArgumentNullException(nameof(trainer));
+
+            if (performanceCounters == null)
+                throw new ArgumentNullException(nameof(performanceCounters));
+
+            this.trainer = trainer;
+            this.performanceCounters = performanceCounters;
+
+            this.telemetry = new TelemetryClient();
+            this.telemetry.Context.Component.Version = GetType().Assembly.GetName().Version.ToString();
+
+            this.evalOperation = new EvalOperation(settings);
+            //this.latencyOperation = new LatencyOperation();
+
+            this.deserializeBlock = new TransformManyBlock<PipelineData, PipelineData>(
+                (Func<PipelineData, IEnumerable<PipelineData>>)this.Deserialize,
+                new ExecutionDataflowBlockOptions
+                {
+                    MaxDegreeOfParallelism = 4, // Math.Max(2, Environment.ProcessorCount - 1),
```

Comment:
should this be turned back on to ProcesscorCount or did you find it troublesome? in the past I found that proc count or proc count * 2 are typically good defaults